source/common/upstream/host_utility.cc

```diff
@@ -196,5 +197,71 @@ HostConstSharedPtr HostUtility::selectOverrideHost(const HostMap* host_map, Host
   return nullptr;
 }
 
+void HostUtility::forEachHostMetric(
+    const ClusterManager& cluster_manager,
+    const std::function<void(Stats::PrimitiveCounterSnapshot&& metric)>& counter_cb,
+    const std::function<void(Stats::PrimitiveGaugeSnapshot&& metric)>& gauge_cb) {
+  for (const auto& [unused_name, cluster_ref] : cluster_manager.clusters().active_clusters_) {
+    if (cluster_ref.get().info()->perEndpointStatsEnabled()) {
+      const std::string cluster_name =
+          Stats::Utility::sanitizeStatsName(cluster_ref.get().info()->observabilityName());
+
+      const Stats::TagVector& fixed_tags =
+          cluster_ref.get().info()->statsScope().store().fixedTags();
+
+      for (auto& host_set : cluster_ref.get().prioritySet().hostSetsPerPriority()) {
+        for (auto& host : host_set->hosts()) {
+
+          Stats::TagVector tags;
+          tags.reserve(fixed_tags.size() + 3);
+          tags.insert(tags.end(), fixed_tags.begin(), fixed_tags.end());
+          tags.emplace_back(Stats::Tag{Envoy::Config::TagNames::get().CLUSTER_NAME, cluster_name});
+          tags.emplace_back(Stats::Tag{"envoy.endpoint_address", host->address()->asString()});
+
+          const auto& hostname = host->hostname();
+          if (!hostname.empty()) {
+            tags.push_back({"envoy.endpoint_hostname", hostname});
```

Comment:
note that if you hoist the tags declaration we'd  need to change the logic here.